### PR TITLE
feat(eslint): create custom rule for deprecated alias

### PR DESCRIPTION
**Pull request by Natalie-Smirnova**
Initial PR for ESlint custom rule to find and replace deprecated aliases

### CODE REVIEW

File: jest.config.js [hunk 1]
```diff
@@ -3,8 +3,8 @@ module.exports = {
     '^.+\\.tsx?$': ['ts-jest']
   },
   testEnvironment: 'jsdom',
-  roots: ['src/modules', 'src/polyfills'],
-  testRegex: '/test/(.+)\\.test\\.ts$',
+  roots: ['eslint'],
```
Comment (ala-n) [inline jest.config.js:6]: I think we shouln't remove existing testing roots out of jest scope.

File: package.json [hunk 2]
```diff
@@ -189,5 +190,12 @@
       "name": "alesun",
       "email": "alesun@exadel.com"
     }
+  ],
+  "eslint": [
+    {
+      "name": "eslint-plugin-custom-rules",
+      "version": "1.0.0",
+      "main": "index.js"
```
Comment (ala-n) [inline package.json:198]: It looks like meta information for lint module is not correct.
"main" should targeting init file for all created lint rules, which is "eslint/index.js"
Regarding "version" I think it have sense to make it equal to actual library version.

File: eslint/index.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+exports.rules = {
```
Comment (ala-n) [inline eslint/index.js:1]: todo: other actual deprecations

File: package.json [hunk 4]
```diff
@@ -103,6 +103,7 @@
     "concurrently": "^7.6.0",
     "copyfiles": "^2.4.1",
     "eslint": "^8.29.0",
+    "eslint-plugin-custom-rules": "file:eslint",
```
Comment (ala-n) [inline package.json:106]:
```suggestion
    "eslint-plugin-esl": "file:eslint",
```

File: eslint/base-rules/deprecated-alias.js [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+ * @param {string} deprecation - deprecated name
+ * @param {string} name - current name
+ */
+
+module.exports.buildRule = function buildRule({alias, deprecation}) {
+  function isDeprecated(node, name, context) {
+    const sourceCode = context.getSourceCode();
+    let currentNode = node.parent;
+    while (currentNode !== null) {
+      if (currentNode.type === "ImportDeclaration") {
```
Comment (dshovchko) [inline eslint/base-rules/deprecated-alias.js:51]:
```suggestion
      if (currentNode.type === 'ImportDeclaration') {
```

File: eslint/base-rules/deprecated-alias.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        return true;
+      } else if (currentNode.type === "Program") {
+        const importStatements = currentNode.body
+          .filter((item) => item.type === "ImportDeclaration")
```
Comment (dshovchko) [inline eslint/base-rules/deprecated-alias.js:55]:
```suggestion
          .filter((item) => item.type === 'ImportDeclaration')
```

File: eslint/base-rules/deprecated-alias.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+ */
+
+module.exports.buildRule = function buildRule({alias, deprecation}) {
+  function isDeprecated(node, name, context) {
+    const sourceCode = context.getSourceCode();
+    let currentNode = node.parent;
+    while (currentNode !== null) {
+      if (currentNode.type === "ImportDeclaration") {
+        return true;
+      } else if (currentNode.type === "Program") {
```
Comment (dshovchko) [inline eslint/base-rules/deprecated-alias.js:53]:
```suggestion
      } else if (currentNode.type === 'Program') {
```

File: eslint/test/deprecated-alias.test.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+// TODO: rewrite for base rule
+"use strict";
```
Comment (dshovchko) [inline eslint/test/deprecated-alias.test.js:2]:
```suggestion
'use strict';
```

File: eslint/test/deprecated-alias.test.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+const rule = require('../rules/4.0.0/deprecated.traversing-query');
+const RuleTester = require("eslint").RuleTester;
```
Comment (dshovchko) [inline eslint/test/deprecated-alias.test.js:5]:
```suggestion
const RuleTester = require('eslint').RuleTester;
```

**Review by dshovchko — APPROVED** [review]

File: eslint/base-rules/deprecated-alias.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+const meta = {
+  type: 'suggestion',
+
+  docs: {
+    description: 'replace deprecated aliases',
+    recommended: true,
+  },
+  settings: {
+    'import/parsers': {
+      '@typescript-eslint/parser': ['.ts', '.js']
```
Comment (ala-n) [inline eslint/base-rules/deprecated-alias.js:13]:
```suggestion
      '@typescript-eslint/parser': ['.ts', '.js', '.tsx', '.jsx']
```

File: eslint/base-rules/deprecated-alias.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * @property {string} alias - current name
+ * @property {string} deprecation - deprecated name
+ */
+
+/**
+ * @param {object} context - element for which to get context
+ * @param {object} node - element for which to get node
+ */
```
Comment (ala-n) [inline eslint/base-rules/deprecated-alias.js:27]:
```suggestion
/**
 * @property {string} alias - current name
 * @property {string} deprecation - deprecated name
 */
```

**Review by fshovchko — APPROVED** [review]

File: eslint/base-rules/deprecated-alias.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+  const uniqNodes = [];
+  for (const node of nodes) {
+    if (!uniqNodes.some((item) => String(item.range) === String(node.range))) {
+      uniqNodes.push(node);
+    }
+  }
+  return uniqNodes.map((node) => node.range);
+}
+
+/** Find the root node in the tree */
```
Comment (NastaLeo) [inline eslint/base-rules/deprecated-alias.js:127]:
```suggestion
/** Finds the root node in the tree */
```

**Review by NastaLeo — APPROVED** [review]

**Review by NastaLeo — APPROVED** [review]

**Comment by ala-n** [comment]
@exadel-inc/esl-core-team, please be aware that this PR is opened against the /epic branch and is going to be merged soon to unlock related epic substories.

**Pull request merged**